packages/medusa/src/api/store/carts/[id]/complete/route.ts

```diff
@@ -1,22 +1,48 @@
 import { completeCartWorkflow } from "@medusajs/core-flows"
-import { MedusaError } from "@medusajs/utils"
+import {
+  MedusaError,
+  ModuleRegistrationName,
+  TransactionState,
+} from "@medusajs/utils"
 import { MedusaRequest, MedusaResponse } from "../../../../../types/routing"
+import { prepareRetrieveQuery } from "../../../../../utils/get-query-config"
+import { refetchEntity } from "../../../../utils/refetch-entity"
 import { refetchOrder } from "../../../orders/helpers"
 import { refetchCart } from "../../helpers"
 import { defaultStoreCartFields } from "../../query-config"
 import { StoreCompleteCartType } from "../../validators"
-import { prepareRetrieveQuery } from "../../../../../utils/get-query-config"
 
 export const POST = async (
   req: MedusaRequest<StoreCompleteCartType>,
   res: MedusaResponse
 ) => {
   const cart_id = req.params.id
 
+  const eventBusModule = req.scope.resolve(ModuleRegistrationName.EVENT_BUS)
+
   const { errors, result } = await completeCartWorkflow(req.scope).run({
     input: { id: cart_id },
     context: { transactionId: cart_id },
     throwOnError: false,
+    events: {
+      onFinish: async ({ transaction }) => {
+        const state = transaction.getState()
+
+        // We only want to emit the order.placed event if the workflow completed
+        if (state !== TransactionState.DONE) {
+          return
+        }
+
+        const result = await refetchEntity(
+          "order_cart",
+          { cart_id },
+          req.scope,
+          ["order_id", "cart_id"]
+        )
+
+        await eventBusModule.emit("order.placed", { id: result.order_id })
+      },
+    },
```

Comment:
@carlos-r-l-rodrigues @adrien2p, what do you think about this?

I kept the idempotency out of this PR. We can address it in a follow up, when we have time to discuss the options.

Reply:
This was added under the assumption that the endpoint `POST /store/carts/:id/complete` would only be used to complete carts and convert them to an order.

If we choose to also handle exchange/swap completion through this endpoint, we will need to rethink the approach.

Reply:
I am not a big fan of the API route owning the event emission. Couldn't this be a step in the workflow instead? 

My thinking is that people might want to build an experience on top of the cart+order flows, (e.g., a quotation flow). This experience will maybe have some different validation than the default cart endpoints, but still rely on the cartCompletionWorkflow. I'd expect `order.placed` to be emitted in this case without having to explicitly add the onFinish handler like this.

Reply:
I don't believe exchange and swap will use this flow. The order edits in general will have their own flows.
In case we want to reuse the cart for some reason to complete these other flows, they will be new carts anyway and they wouldn't generate new orders

Reply:
But I think you can use the workflow engine here and use the result to send the event. The idempotency part will come in later

Reply:
> Couldn't this be a step in the workflow instead?

The reason I didn't add it as a step is that we currently can't guarantee it is the last step in the workflow. As mentioned in a separate comment, if we add controlled extensions, this is reasonable and maybe even the preferred approach.

Should we instead introduce this functionality under the assumption, that we will do that in the near future?

Reply:
When we have events grouped, even if the event is not the last step, it will only be fired when the flow is completed.